synapse_spamcheck_badlist/BadListFilter.py

```diff
@@ -0,0 +1,212 @@
+                for match in self._linkifier.match(text.lower()) or []:
+                    link = re.sub(self._scheme_re, "", match.url)
+                    is_bad_link = await self._api.run_db_interaction("Check link against evil db", _db_is_bad_link, self._links_table, link)
+                    if is_bad_link:
+                        logger.info("Rejected bad link")
+                        return True
+
+        # If it's a file, download content, extract hash.
+        elif event["content"]["msgtype"] == "m.file":
+            if not await self.can_we_check_md5():
+                return False
+
+            match = self._mxc_re.match(event["content"]["url"])
+            server_name = None
+            media_id = None
+            if match != None:
+                server_name = match.group('server_name')
+                media_id = match.group('media_id')
+                response = None
+                try:
+                    url = "%s/_matrix/media/r0/download/%s/%s" % (
+                            self._base_url,
+                            urlquote(server_name),
+                            urlquote(media_id)
+                    )
+                    response = await self._api.http_client.request("GET", url)
+                    # FIXME: What do we do if we were rate-limited?
+                except Exception as e:
+                    # In case of timeout or error, there's nothing we can do.
+                    # Let's not take the risk of blocking valid contents.
+                    logger.warn("Could not download media: %s" % e)
+                md5 = hashlib.md5()
+                await response.collect(lambda batch: md5.update(batch))
+                is_bad_upload = await self._api.run_db_interaction("Check upload against evil db", _db_is_bad_upload, self._md5_table, md5.hexdigest())
+                if is_bad_upload:
+                    logger.info("Rejected bad upload")
+                    return True
+
+        # Not spam
+        return False
+
+    def user_may_invite(
+        self, inviter_userid: str, invitee_userid: str, room_id: str
+    ) -> bool:
+        return True
+
+    def user_may_create_room(self, userid: str) -> bool:
+        return True
+
+    def user_may_create_room_alias(self, userid: str, room_alias: str) -> bool:
+        return True
+
+    def user_may_publish_room(self, userid: str, room_id: str) -> bool:
+        return True
+
+    @staticmethod
+    def parse_config(config):
+        # No parsing needed
+        return config
+
+
+def _db_is_bad_link(db, table, link):
+    """
+    Search if any url in the database is a prefix of `link`.
+    `link` MUST be normalized by `_link_for_search`.
+    """
+    # Note: As per IWF guidelines, we're looking for *prefixes*. This might
+    # be slow. We're quickly going to have 1M+ links, so we need to find out
+    # whether this slows things down.
```

Comment:
I think you and @richvdh discussed this query so I'm going to believe that this works. 😄